src/foam/core/CompoundContextAgent.java

```diff
@@ -0,0 +1,66 @@
+/**
+ * @license
+ * Copyright 2019 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.core;
+
+import java.util.ArrayList;
+
+/**
+ * A ContextAgent which is itself an Agency.
+ * When ContextAgents are submit()'ed to the service it queues them
+ * but doesn't execute them until it is itself executed.
+ **/
+public class CompoundContextAgent
+  implements ContextAgent, Agency
+{
+  protected ArrayList<Runnable> agents_ = new ArrayList();
+  protected X x_;
+  public CompoundContextAgent(X x) {
+    x_ = x;
+  }
+
+  public void submit(ContextAgent agent, String description) {
+    agents_.add(new ContextAgentRunnable(x_, agent, description));
+  }
+
+  public void submit(X x, ContextAgent agent) {
+    submit(agent, "");
+  }
+
+  public void execute(X x) {
+    for ( Runnable agent : agents_ ) {
+      try {
+        agent.run();
+      } catch (Throwable t) {
+      }
```

Comment:
The next agent will still be run if the previous one throws. Is it something we want?

Reply:
the idea was that agent should not throw anything ... by the time we execute agents, we "know" that all the actions can be successfully executed, otherwise there is no point in them.